File: 110-balanced-binary-tree/balanced-binary-tree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isBalanced(_ root: TreeNode?) -> Bool {
        func balanced(current node: TreeNode?) -> (Int, Bool){
            guard let curr = node else {return (0, true)};
            var left: (Int, Bool) = balanced(current: curr.left);
            if !left.1{
                return (0, false);
            }
            var right: (Int, Bool) = balanced(current: curr.right);
            if !right.1{
                return (0, false);
            }

            if abs(left.0 - right.0) > 1{
                return (0, false);
            }

            return (max(left.0, right.0) + 1, true)
        }

        return balanced(current: root).1;
    }
}

File: 110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def re(node):
            if not node:
                return [0, True]

            left = re(node.left)
            right = re(node.right)

            if not left[1] or not right[1]:
                return [0, False]
            return [max(left[0] , right[0]) + 1, abs(left[0] - right[0]) <= 1]

        return re(root)[1]
        